# Capstone Project - The Battle of the Neighborhoods 
### Applied Data Science Capstone by IBM/Coursera

## Introduction: Problem Description 

Lagos and Accra two of West Africa's commercial cities located in Nigeria and Ghana respectively. The city of Lagos is a major economic focal point in Nigeria as well Accra, Ghana which also serves as the country's capital. With as many things in common as Nigerians and Ghanaians have, there are still many things citizens of either countries still agree and disagree on. Lagos and Accra are both dynamic cities pulsating with creative energy and buzz. There has been ongoing talks, arguments and discussions regarding the quality of life these two cities have to offer its residents in the areas of entertainment, recreational activities, sport centres etc. They are both entrepreneurial hubs, evolving fast, yet still holding on to their own very distinct identities.

This work focuses on analyzing the two neighborhoods earlier mentioned and helps to understand what is popular in them and what they have to offer to someone who is considering making a choice about seeking a life in either of the commercial cities.

### DATA

For this project, we needed data about the neighborhoods in each of these commercial cities. Lagos is divided into 20 Local governemnt area (LGA) 4 among those 20 LGAs are classified under non-metropolitan Lagos State, while the other 16 make up the city of Lagos. Data regarding this information about lagos is found on - [Nigeria Postcode](https://postcode.com.ng/lagos-zip-code-postal-code/). Accra is divided in planned, unplanned and slums neighborhood the dataset for the city of accra is found on - [Neighborhoods of Accra](https://en.wikipedia.org/wiki/Neighborhoods_of_Accra).

For the data set of this project i would extract and clean up the data available on [Nigeria Postcode](https://postcode.com.ng/lagos-zip-code-postal-code/) and [Neighborhoods of Accra](https://en.wikipedia.org/wiki/Neighborhoods_of_Accra) into excel format of rows and columns and upload into pandas dataframe on jupyter notebook. The column Towns in Lagos is changed to Neighborbood in excel. 

Foursquare API will be used to find the longitude and latitude of each of the neighborhoods in both Lagos and Accra. This will form the dataset we will use for this study. Foursquare API would be used to explore the neighborhoods and segment them. 

### METHODOLOGY

We would start by installing Geocoder and importing the required python libraries and packages 

In [2]:
pip install geocoder

Note: you may need to restart the kernel to use updated packages.


In [3]:
import numpy as np 
import pandas as pd 
import json
from geopy.geocoders import Nominatim 
import geocoder # to get coordinates
import requests
from pandas.io.json import json_normalize 
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
import folium 

print('Libraries imported.')

Libraries imported.


For the dataset for Lagos i extracted the towns/cities(Neighborhoods) from [Nigeria Postcode](https://postcode.com.ng/lagos-zip-code-postal-code/) into excel format i found it more convenient than using beautifull soup

In [4]:
lagos_data = pd.read_excel(r"file:///C:\Users\User\Downloads\Lagos%20Post%20code.xlsx") 
lagos_data.head(12)

Neighborhood  Postcode                 Area 
0             Abarenji     100269  Metropolitan Lagos 
1           Abule Egba     100276  Metropolitan Lagos 
2         Abule Ijesha     100213  Metropolitan Lagos 
3          Abule Okuta     100223  Metropolitan Lagos 
4        Adeniji Adele     101226  Metropolitan Lagos 
5   Adeniran Ogunsanya     101211  Metropolitan Lagos 
6              Agarawu     101229  Metropolitan Lagos 
7           Agbelekale     100324  Metropolitan Lagos 
8                Agege     100283  Metropolitan Lagos 
9                Agodo     100265  Metropolitan Lagos 
10             Ahmadia     100331  Metropolitan Lagos 
11                Ajah     101245  Metropolitan Lagos

In [5]:
lagos_data.shape

(161, 3)

We currently have 161 rows and 3 columns of our data

I Used geopy library to get the latitude and longitude values of Lagos Nigeria, We will use the Nominatim library from geocoders.geopy package to find the longitude and latitude for each of the neighborhoods and would create a dataset having all the necessary columns for our analysis.

In [6]:
# define a function to get coordinates
def get_latlng(neighborhood):
    # initialize your variable to None
    lat_lng_coords = None
    # loop until you get the coordinates
    while(lat_lng_coords is None):
        g = geocoder.arcgis('{}, Lagos, Nigeria'.format(neighborhood))
        lat_lng_coords = g.latlng
    return lat_lng_coords

In [8]:
# call the function to get the coordinates, store in a new list using list comprehension
coords = [ get_latlng(neighborhood) for neighborhood in lagos_data["Neighborhood"].tolist() ]

In [9]:
coords

[[6.454700000000059, 3.3887600000000475],
 [6.647500000000036, 3.303060000000073],
 [6.524720000000059, 3.3863900000000626],
 [6.539930000000027, 3.3928000000000225],
 [6.464517767877055, 3.3953955859622393],
 [6.494346121698933, 3.3573148928611607],
 [6.4569056364811495, 3.3900216180466587],
 [6.640280000000075, 3.2869400000000724],
 [6.6256100000000515, 3.312620000000038],
 [6.678881112624563, 3.346879869742418],
 [6.454700000000059, 3.3887600000000475],
 [6.434440950953698, 3.5891929806263025],
 [6.604690000000062, 3.3890400000000227],
 [6.603110000000072, 3.285250000000076],
 [6.536960000000022, 3.223220000000026],
 [6.5245664787103665, 3.354689139634434],
 [7.600000000000023, 3.183330000000069],
 [6.456280000000049, 3.2708900000000654],
 [8.616670000000056, 4.433330000000069],
 [6.634830000000022, 3.880520000000047],
 [6.609270000000038, 3.255800000000022],
 [6.5064829730445, 3.3619355179119728],
 [6.466590000000053, 3.3458700000000476],
 [7.216670000000022, 4.250000000000057],
 [

In [11]:
df_coords = pd.DataFrame(coords, columns=['Latitude', 'Longitude'])

In [12]:
# merge the coordinates into the original dataframe
lagos_data['Latitude'] = df_coords['Latitude']
lagos_data['Longitude'] = df_coords['Longitude']

In [13]:
print(lagos_data.shape)
lagos_data.head(12)

(161, 5)


Neighborhood  Postcode                 Area   Latitude  Longitude
0             Abarenji     100269  Metropolitan Lagos   6.454700   3.388760
1           Abule Egba     100276  Metropolitan Lagos   6.647500   3.303060
2         Abule Ijesha     100213  Metropolitan Lagos   6.524720   3.386390
3          Abule Okuta     100223  Metropolitan Lagos   6.539930   3.392800
4        Adeniji Adele     101226  Metropolitan Lagos   6.464518   3.395396
5   Adeniran Ogunsanya     101211  Metropolitan Lagos   6.494346   3.357315
6              Agarawu     101229  Metropolitan Lagos   6.456906   3.390022
7           Agbelekale     100324  Metropolitan Lagos   6.640280   3.286940
8                Agege     100283  Metropolitan Lagos   6.625610   3.312620
9                Agodo     100265  Metropolitan Lagos   6.678881   3.346880
10             Ahmadia     100331  Metropolitan Lagos   6.454700   3.388760
11                Ajah     101245  Metropolitan Lagos   6.434441   3.589193

In [14]:
address = 'Lagos, Nigeria'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Lagos are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Lagos are 6.4550575, 3.3941795.


We now have the necessary information to visualize the neighborhoods of Lagos on folium map

In [15]:
map_lagos = folium.Map(location=[latitude, longitude], zoom_start=11)

In [17]:
# add markers to map
for lat, lng, label in zip(lagos_data['Latitude'], lagos_data['Longitude'], lagos_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_lagos)  
    
map_lagos

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them

In [18]:
CLIENT_ID = 'GNTVLV32ERNVZZCU2TDUNJ2XMWO3QPOXD44HVJLR5RHW1EBC' # your Foursquare ID
CLIENT_SECRET = 'GVISMW05IPDBYCKTX11BBRY41CDIJZOFZLAEHYABACEPKJ3H' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: GNTVLV32ERNVZZCU2TDUNJ2XMWO3QPOXD44HVJLR5RHW1EBC
CLIENT_SECRET:GVISMW05IPDBYCKTX11BBRY41CDIJZOFZLAEHYABACEPKJ3H


In [19]:
lagos_data

Neighborhood  Postcode                    Area   Latitude  \
0                 Abarenji     100269     Metropolitan Lagos   6.454700   
1               Abule Egba     100276     Metropolitan Lagos   6.647500   
2             Abule Ijesha     100213     Metropolitan Lagos   6.524720   
3              Abule Okuta     100223     Metropolitan Lagos   6.539930   
4            Adeniji Adele     101226     Metropolitan Lagos   6.464518   
..                     ...        ...                     ...       ...   
156                 Erodo      106103  Non-Metropolitan Lagos  6.640680   
157                 Ibeju      105101  Non-Metropolitan Lagos  6.487010   
158                  Lekki     105102  Non-Metropolitan Lagos  6.409480   
159  Ikorodu rural areas       104101  Non-Metropolitan Lagos  6.623560   
160              Irepodun      104102  Non-Metropolitan Lagos  6.683795   

     Longitude  
0     3.388760  
1     3.303060  
2     3.386390  
3     3.392800  
4     3.395396  
..         ...  
156   3.990410  
157   3.892810  
158   4.091520  
159   3.504830  
160   3.308795  

[161 rows x 5 columns]

Now, let's get the top 100 venues that are in Lagos within a radius of 2000 meters

In [21]:
radius = 2000
LIMIT = 100

venues = []

for lat, long, neighborhood in zip(lagos_data['Latitude'], lagos_data['Longitude'], lagos_data['Neighborhood']):
    
    # create the API request URL
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    # make the GET request
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    # return only relevant information for each nearby venue
    for venue in results:
        venues.append((
            neighborhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

In [22]:
# convert the venues list into a new DataFrame
venues_df = pd.DataFrame(venues)

# define the column names
venues_df.columns = ['Neighborhood', 'Latitude', 'Longitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']

print(venues_df.shape)
venues_df

(2228, 7)


Neighborhood  Latitude  Longitude  \
0                  Abarenji  6.454700   3.388760   
1                  Abarenji  6.454700   3.388760   
2                  Abarenji  6.454700   3.388760   
3                  Abarenji  6.454700   3.388760   
4                  Abarenji  6.454700   3.388760   
...                     ...       ...        ...   
2223  Ikorodu rural areas    6.623560   3.504830   
2224  Ikorodu rural areas    6.623560   3.504830   
2225  Ikorodu rural areas    6.623560   3.504830   
2226              Irepodun   6.683795   3.308795   
2227              Irepodun   6.683795   3.308795   

                                  VenueName  VenueLatitude  VenueLongitude  \
0                       Campos Mini Stadium       6.451084        3.390552   
1                              Freedom Park       6.449065        3.396536   
2                           SpicyInc Studio       6.451069        3.398125   
3                            Campos Stadium       6.451098        3.394655   
4                           Sweet Sensation       6.453180        3.391035   
...                                     ...            ...             ...   
2223  Tipper Garage,Shagamu Exp Way,Ikorodu       6.630069        3.508764   
2224        Moye Hotel,Shagamu Road,Ikorodu       6.635063        3.516400   
2225                                Club 24       6.612952        3.491228   
2226                            01 Shawarma       6.679488        3.299837   
2227                    Rap hill tv studios       6.684772        3.324660   

             VenueCategory  
0         Baseball Stadium  
1                     Park  
2       Photography Studio  
3             Soccer Field  
4     Fast Food Restaurant  
...                    ...  
2223         Auto Workshop  
2224                 Hotel  
2225                   Bar  
2226           Snack Place  
2227           Music Venue  

[2228 rows x 7 columns]

In [23]:
venues_df.groupby(["Neighborhood"]).count()

Latitude  Longitude  VenueName  VenueLatitude  \
Neighborhood                                                           
Abarenji                     22         22         22             22   
Abule Egba                    5          5          5              5   
Abule Ijesha                 16         16         16             16   
Abule Okuta                  10         10         10             10   
Adeniji Adele                22         22         22             22   
...                         ...        ...        ...            ...   
Suberu Oje                    3          3          3              3   
Surulere Central             30         30         30             30   
Tujusho                      22         22         22             22   
Victoria Island             100        100        100            100   
Yaba/ Ebute Meta East         9          9          9              9   

                       VenueLongitude  VenueCategory  
Neighborhood                                          
Abarenji                           22             22  
Abule Egba                          5              5  
Abule Ijesha                       16             16  
Abule Okuta                        10             10  
Adeniji Adele                      22             22  
...                               ...            ...  
Suberu Oje                          3              3  
Surulere Central                   30             30  
Tujusho                            22             22  
Victoria Island                   100            100  
Yaba/ Ebute Meta East               9              9  

[144 rows x 6 columns]

In [24]:
# print out the list of categories
venues_df['VenueCategory'].unique()[:60]

array(['Baseball Stadium', 'Park', 'Photography Studio', 'Soccer Field',
       'Fast Food Restaurant', 'Convention Center', 'Lounge', 'Plaza',
       'Fried Chicken Joint', 'Market', 'Shopping Mall', 'Pizza Place',
       'Bank', 'Pharmacy', 'Airport', 'Harbor / Marina', 'Flea Market',
       'Train Station', 'Burger Joint', 'Hookah Bar', 'Farmers Market',
       'Light Rail Station', 'Bakery', 'Bagel Shop', 'Department Store',
       'Nightclub', 'Arts & Entertainment', 'Bus Station',
       'Convenience Store', 'Boat or Ferry', 'Breakfast Spot',
       'Multiplex', 'Movie Theater', 'African Restaurant',
       'Clothing Store', 'Cupcake Shop', 'Art Gallery', 'BBQ Joint',
       'Stadium', 'Restaurant', 'Jewelry Store',
       'Bike Rental / Bike Share', 'Seafood Restaurant', 'Bus Stop',
       'IT Services', 'Campground', 'Gym', 'Beach', 'Hot Dog Joint',
       'Hospital', 'Hotel', 'Boutique', 'Basketball Court', 'Music Venue',
       'Airport Terminal', 'Food Court', 'Rental Car Lo

In [25]:
print('There are {} uniques categories.'.format(len(venues_df['VenueCategory'].unique())))

There are 153 uniques categories.


In [26]:
"Neighborhood" in venues_df['VenueCategory'].unique()

True

In [27]:
# one hot encoding
lagos_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
lagos_onehot['Neighborhoods'] = venues_df['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [lagos_onehot.columns[-1]] + list(lagos_onehot.columns[:-1])
lagos_onehot = lagos_onehot[fixed_columns]

print(lagos_onehot.shape)
lagos_onehot.head()

(2228, 154)


Neighborhoods  Accessories Store  African Restaurant  Airport  \
0      Abarenji                  0                   0        0   
1      Abarenji                  0                   0        0   
2      Abarenji                  0                   0        0   
3      Abarenji                  0                   0        0   
4      Abarenji                  0                   0        0   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0               0                0                 0   
1                   0               0                0                 0   
2                   0               0                0                 0   
3                   0               0                0                 0   
4                   0               0                0                 0   

   American Restaurant  Arcade  ...  Theater  Theme Restaurant  Track  Trail  \
0                    0       0  ...        0                 0      0      0   
1                    0       0  ...        0                 0      0      0   
2                    0       0  ...        0                 0      0      0   
3                    0       0  ...        0                 0      0      0   
4                    0       0  ...        0                 0      0      0   

   Train Station  Vietnamese Restaurant  Vineyard  Water Park  Wine Bar  \
0              0                      0         0           0         0   
1              0                      0         0           0         0   
2              0                      0         0           0         0   
3              0                      0         0           0         0   
4              0                      0         0           0         0   

   Wine Shop  
0          0  
1          0  
2          0  
3          0  
4          0  

[5 rows x 154 columns]

In [65]:

lagos_grouped = lagos_onehot.groupby(["Neighborhoods"]).mean().reset_index()

print(lagos_grouped.shape)
lagos_grouped

(144, 154)


Neighborhoods  Accessories Store  African Restaurant   Airport  \
0                 Abarenji                0.0            0.000000  0.045455   
1               Abule Egba                0.0            0.000000  0.000000   
2             Abule Ijesha                0.0            0.000000  0.000000   
3              Abule Okuta                0.0            0.000000  0.000000   
4            Adeniji Adele                0.0            0.000000  0.045455   
..                     ...                ...                 ...       ...   
139             Suberu Oje                0.0            0.000000  0.000000   
140       Surulere Central                0.0            0.033333  0.000000   
141                Tujusho                0.0            0.000000  0.045455   
142        Victoria Island                0.0            0.060000  0.000000   
143  Yaba/ Ebute Meta East                0.0            0.000000  0.000000   

     Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0.0             0.0              0.0               0.0   
1                   0.0             0.0              0.0               0.0   
2                   0.0             0.0              0.0               0.0   
3                   0.0             0.0              0.0               0.0   
4                   0.0             0.0              0.0               0.0   
..                  ...             ...              ...               ...   
139                 0.0             0.0              0.0               0.0   
140                 0.0             0.0              0.0               0.0   
141                 0.0             0.0              0.0               0.0   
142                 0.0             0.0              0.0               0.0   
143                 0.0             0.0              0.0               0.0   

     American Restaurant  Arcade  ...  Theater  Theme Restaurant  Track  \
0                    0.0     0.0  ...     0.00               0.0    0.0   
1                    0.0     0.0  ...     0.00               0.0    0.0   
2                    0.0     0.0  ...     0.00               0.0    0.0   
3                    0.0     0.0  ...     0.00               0.0    0.0   
4                    0.0     0.0  ...     0.00               0.0    0.0   
..                   ...     ...  ...      ...               ...    ...   
139                  0.0     0.0  ...     0.00               0.0    0.0   
140                  0.0     0.0  ...     0.00               0.0    0.0   
141                  0.0     0.0  ...     0.00               0.0    0.0   
142                  0.0     0.0  ...     0.01               0.0    0.0   
143                  0.0     0.0  ...     0.00               0.0    0.0   

     Trail  Train Station  Vietnamese Restaurant  Vineyard  Water Park  \
0      0.0       0.045455                    0.0       0.0         0.0   
1      0.0       0.000000                    0.0       0.0         0.0   
2      0.0       0.000000                    0.0       0.0         0.0   
3      0.0       0.000000                    0.0       0.0         0.0   
4      0.0       0.045455                    0.0       0.0         0.0   
..     ...            ...                    ...       ...         ...   
139    0.0       0.000000                    0.0       0.0         0.0   
140    0.0       0.000000                    0.0       0.0         0.0   
141    0.0       0.045455                    0.0       0.0         0.0   
142    0.0       0.000000                    0.0       0.0         0.0   
143    0.0       0.111111                    0.0       0.0         0.0   

     Wine Bar  Wine Shop  
0         0.0        0.0  
1         0.0        0.0  
2         0.0        0.0  
3         0.0        0.0  
4         0.0        0.0  
..        ...        ...  
139       0.0        0.0  
140       0.0        0.0  
141       0.0        0.0  
142       0.0        0.0  
143       0.0        0.0  

[144 rows x 

In [66]:
lagos_grouped.shape

(144, 154)

In [67]:
lagos_grouped = lagos_onehot.groupby('Neighborhoods').mean().reset_index()
lagos_grouped

Neighborhoods  Accessories Store  African Restaurant   Airport  \
0                 Abarenji                0.0            0.000000  0.045455   
1               Abule Egba                0.0            0.000000  0.000000   
2             Abule Ijesha                0.0            0.000000  0.000000   
3              Abule Okuta                0.0            0.000000  0.000000   
4            Adeniji Adele                0.0            0.000000  0.045455   
..                     ...                ...                 ...       ...   
139             Suberu Oje                0.0            0.000000  0.000000   
140       Surulere Central                0.0            0.033333  0.000000   
141                Tujusho                0.0            0.000000  0.045455   
142        Victoria Island                0.0            0.060000  0.000000   
143  Yaba/ Ebute Meta East                0.0            0.000000  0.000000   

     Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0.0             0.0              0.0               0.0   
1                   0.0             0.0              0.0               0.0   
2                   0.0             0.0              0.0               0.0   
3                   0.0             0.0              0.0               0.0   
4                   0.0             0.0              0.0               0.0   
..                  ...             ...              ...               ...   
139                 0.0             0.0              0.0               0.0   
140                 0.0             0.0              0.0               0.0   
141                 0.0             0.0              0.0               0.0   
142                 0.0             0.0              0.0               0.0   
143                 0.0             0.0              0.0               0.0   

     American Restaurant  Arcade  ...  Theater  Theme Restaurant  Track  \
0                    0.0     0.0  ...     0.00               0.0    0.0   
1                    0.0     0.0  ...     0.00               0.0    0.0   
2                    0.0     0.0  ...     0.00               0.0    0.0   
3                    0.0     0.0  ...     0.00               0.0    0.0   
4                    0.0     0.0  ...     0.00               0.0    0.0   
..                   ...     ...  ...      ...               ...    ...   
139                  0.0     0.0  ...     0.00               0.0    0.0   
140                  0.0     0.0  ...     0.00               0.0    0.0   
141                  0.0     0.0  ...     0.00               0.0    0.0   
142                  0.0     0.0  ...     0.01               0.0    0.0   
143                  0.0     0.0  ...     0.00               0.0    0.0   

     Trail  Train Station  Vietnamese Restaurant  Vineyard  Water Park  \
0      0.0       0.045455                    0.0       0.0         0.0   
1      0.0       0.000000                    0.0       0.0         0.0   
2      0.0       0.000000                    0.0       0.0         0.0   
3      0.0       0.000000                    0.0       0.0         0.0   
4      0.0       0.045455                    0.0       0.0         0.0   
..     ...            ...                    ...       ...         ...   
139    0.0       0.000000                    0.0       0.0         0.0   
140    0.0       0.000000                    0.0       0.0         0.0   
141    0.0       0.045455                    0.0       0.0         0.0   
142    0.0       0.000000                    0.0       0.0         0.0   
143    0.0       0.111111                    0.0       0.0         0.0   

     Wine Bar  Wine Shop  
0         0.0        0.0  
1         0.0        0.0  
2         0.0        0.0  
3         0.0        0.0  
4         0.0        0.0  
..        ...        ...  
139       0.0        0.0  
140       0.0        0.0  
141       0.0        0.0  
142       0.0        0.0  
143       0.0        0.0  

[144 rows x 

In [68]:
num_top_venues = 5

for hood in lagos_grouped['Neighborhoods']:
    print("----" +hood+ "----")
    temp = lagos_grouped[lagos_grouped['Neighborhoods'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Abarenji----
                  venue  freq
0  Fast Food Restaurant  0.14
1                Market  0.09
2       Harbor / Marina  0.09
3   Fried Chicken Joint  0.05
4           Pizza Place  0.05


----Abule Egba----
                venue  freq
0         Flea Market   0.2
1  Light Rail Station   0.2
2          Hookah Bar   0.2
3      Farmers Market   0.2
4        Burger Joint   0.2


----Abule Ijesha----
                  venue  freq
0  Fast Food Restaurant  0.19
1     Convenience Store  0.06
2              Pharmacy  0.06
3         Boat or Ferry  0.06
4             Nightclub  0.06


----Abule Okuta----
                  venue  freq
0      Department Store   0.2
1           Bus Station   0.2
2  Fast Food Restaurant   0.2
3                Lounge   0.1
4     Convenience Store   0.1


----Adeniji Adele----
                  venue  freq
0                  Park  0.14
1  Fast Food Restaurant  0.14
2                Market  0.09
3          Soccer Field  0.09
4       Harbor / Marina  0.09


---

In [69]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [70]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhoods']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhoods'] = lagos_grouped['Neighborhoods']

for ind in np.arange(lagos_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(lagos_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhoods 1st Most Common Venue 2nd Most Common Venue  \
0                 Abarenji  Fast Food Restaurant       Harbor / Marina   
1               Abule Egba            Hookah Bar        Farmers Market   
2             Abule Ijesha  Fast Food Restaurant                  Park   
3              Abule Okuta  Fast Food Restaurant           Bus Station   
4            Adeniji Adele                  Park  Fast Food Restaurant   
..                     ...                   ...                   ...   
139             Suberu Oje   Sporting Goods Shop                   Gym   
140       Surulere Central         Shopping Mall  Fast Food Restaurant   
141                Tujusho  Fast Food Restaurant       Harbor / Marina   
142        Victoria Island                 Hotel    Italian Restaurant   
143  Yaba/ Ebute Meta East                  Bank              Pharmacy   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                  Market           Flea Market              Pharmacy   
1      Light Rail Station          Burger Joint           Flea Market   
2             Pizza Place                Market     Convenience Store   
3        Department Store           Pizza Place         Boat or Ferry   
4         Harbor / Marina                Market          Soccer Field   
..                    ...                   ...                   ...   
139          Burger Joint             Wine Shop          Dessert Shop   
140        Clothing Store             Nightclub         Movie Theater   
141                Market           Flea Market              Pharmacy   
142    African Restaurant                Lounge                  Café   
143           Bus Station          Liquor Store  Fast Food Restaurant   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0      Photography Studio           Pizza Place                 Plaza   
1                   Diner           Fish Market                 Field   
2               Nightclub              Pharmacy           Bus Station   
3                  Lounge     Convenience Store                 Field   
4           Train Station           Flea Market    Photography Studio   
..                    ...                   ...                   ...   
139                 Field  Fast Food Restaurant        Farmers Market   
140             BBQ Joint    Chinese Restaurant             Cafeteria   
141    Photography Studio           Pizza Place                 Plaza   
142             Nightclub     Indian Restaurant            Restaurant   
143      Basketball Court         Train Station     Convenience Store   

    9th Most Common Venue 10th Most Common Venue  
0       Convention Center                 Lounge  
1    Fast Food Restaurant      Electronics Store  
2        Department Store   Arts & Entertainment  
3          Farmers Market      Electronics Store  
4            Burger Joint          Boat or Ferry  
..                    ...                    ...  
139     Electronics Store                Dog Run  
140           Bus Station           Cupcake Shop  
141     Convention Center                 Lounge  
142        Ice Cream Shop                    Bar  
143     Convention Center           Cupcake Shop  

[144 rows x 11 columns]

<h3>Cluster the Neighborhoods</h3> 

In [71]:
# set number of clusters
kclusters = 5

lagos_grouped_clustering = lagos_grouped.drop('Neighborhoods', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(lagos_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([3, 4, 3, 1, 3, 3, 3, 1, 1, 3])

In [72]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

lagos_merged = lagos_data

# merge toronto_grouped with lagos_data to add latitude/longitude for each neighborhood
lagos_merged = lagos_merged.join(neighborhoods_venues_sorted.set_index('Neighborhoods'), on='Neighborhood')

lagos_merged.head() # check the last columns!

Neighborhood  Postcode                 Area   Latitude  Longitude  \
0       Abarenji     100269  Metropolitan Lagos   6.454700   3.388760   
1     Abule Egba     100276  Metropolitan Lagos   6.647500   3.303060   
2   Abule Ijesha     100213  Metropolitan Lagos   6.524720   3.386390   
3    Abule Okuta     100223  Metropolitan Lagos   6.539930   3.392800   
4  Adeniji Adele     101226  Metropolitan Lagos   6.464518   3.395396   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0             3.0  Fast Food Restaurant       Harbor / Marina   
1             4.0            Hookah Bar        Farmers Market   
2             3.0  Fast Food Restaurant                  Park   
3             1.0  Fast Food Restaurant           Bus Station   
4             3.0                  Park  Fast Food Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                Market           Flea Market              Pharmacy   
1    Light Rail Station          Burger Joint           Flea Market   
2           Pizza Place                Market     Convenience Store   
3      Department Store           Pizza Place         Boat or Ferry   
4       Harbor / Marina                Market          Soccer Field   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0    Photography Studio           Pizza Place                 Plaza   
1                 Diner           Fish Market                 Field   
2             Nightclub              Pharmacy           Bus Station   
3                Lounge     Convenience Store                 Field   
4         Train Station           Flea Market    Photography Studio   

  9th Most Common Venue 10th Most Common Venue  
0     Convention Center                 Lounge  
1  Fast Food Restaurant      Electronics Store  
2      Department Store   Arts & Entertainment  
3        Farmers Market      Electronics Store  
4          Burger Joint          Boat or Ferry

In [73]:
lagos_merged = lagos_merged.dropna()

In [75]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(lagos_merged['Latitude'], lagos_merged['Longitude'], lagos_merged['Neighborhood'], lagos_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster) - 1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [76]:
lagos_merged.loc[lagos_merged['Cluster Labels'] == 0]

Neighborhood  Postcode                    Area   Latitude  Longitude  \
47      Epe Tedo     101224     Metropolitan Lagos    6.58375   3.975530   
53         G.R.A     100271     Metropolitan Lagos    6.58477   3.986005   
155         Epe      106101  Non-Metropolitan Lagos   6.58375   3.975530   

     Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
47              0.0                 Hotel             Wine Shop   
53              0.0                 Hotel             Wine Shop   
155             0.0                 Hotel             Wine Shop   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
47           Dessert Shop           Fish Market                 Field   
53           Dessert Shop           Fish Market                 Field   
155          Dessert Shop           Fish Market                 Field   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
47   Fast Food Restaurant        Farmers Market     Electronics Store   
53   Fast Food Restaurant        Farmers Market     Electronics Store   
155  Fast Food Restaurant        Farmers Market     Electronics Store   

    9th Most Common Venue 10th Most Common Venue  
47                Dog Run                  Diner  
53                Dog Run                  Diner  
155               Dog Run                  Diner

In [77]:
lagos_merged.loc[lagos_merged['Cluster Labels'] == 1]

Neighborhood  Postcode                    Area   Latitude  \
3                Abule Okuta     100223     Metropolitan Lagos   6.539930   
7                 Agbelekale     100324     Metropolitan Lagos   6.640280   
8                      Agege     100283     Metropolitan Lagos   6.625610   
12               Ajao Estate     100263     Metropolitan Lagos   6.604690   
14                    Akesan     100268     Metropolitan Lagos   6.536960   
15                   Akintan     100341     Metropolitan Lagos   6.524566   
20                  Alimosho     100275     Metropolitan Lagos   6.609270   
22         Amukoko Alaba Oro     102212     Metropolitan Lagos   6.466590   
31   Atunrase Estate Gbagada     100235     Metropolitan Lagos   6.549490   
36               Bode Thomas     101211     Metropolitan Lagos   6.535176   
41                    Dopemu     100273     Metropolitan Lagos   6.617219   
42                   Dosunmu     101229     Metropolitan Lagos   6.514910   
45                      Egbe     100265     Metropolitan Lagos   6.543890   
48                    Fadeyi     100252     Metropolitan Lagos   6.524952   
54                   Gbagada     100234     Metropolitan Lagos   6.544155   
55                 Idi Araba     100254     Metropolitan Lagos   6.520280   
56                   Idi Oro     100254     Metropolitan Lagos   6.522500   
57                  Idumagbo     101228     Metropolitan Lagos   6.528705   
58                   Idumota     101227     Metropolitan Lagos   6.535465   
60               Ifako Agege     100215     Metropolitan Lagos   6.625610   
62                     Ijaye     100311     Metropolitan Lagos   6.681879   
72                 Iju Isaga     100216     Metropolitan Lagos   6.675000   
73           Iju Water Works     100216     Metropolitan Lagos   6.673460   
77                    Ikotun     100265     Metropolitan Lagos   6.543890   
79                     Ilaje     102266     Metropolitan Lagos   6.608260   
80                   Ilupeju     100252     Metropolitan Lagos   6.539846   
84                Isheri Oke     100214     Metropolitan Lagos   6.640043   
86                     Isolo     100263     Metropolitan Lagos   6.529160   
93             Ketu Orisigun     100245     Metropolitan Lagos   6.601390   
94       Kirikiri Industrial     102272     Metropolitan Lagos   6.443060   
95            Kirikiri Phase     102273     Metropolitan Lagos   6.443060   
97                  Lawanson     100254     Metropolitan Lagos   6.512746   
106                   Mushin     100253     Metropolitan Lagos   6.531740   
109                    Offin     101227     Metropolitan Lagos   6.545620   
110               Ogba Aguda     100218     Metropolitan Lagos   6.634720   
112                    Ojodu     100213     Metropolitan Lagos   6.666505   
114                    Ojota     100242     Metropolitan Lagos   6.586310   
115                 Oke Arin     101227     Metropolitan Lagos   6.538814   
117            Oko Oba Agege     100284     Metropolitan Lagos   6.639010   
119                    Olota     100341     Metropolitan Lagos   6.586310   
125                  Onipanu     100252     Metropolitan Lagos   6.532720   
131      Oworonshoki L and K     100222     Metropolitan Lagos   6.545940   
140          Shomolu Central     100231     Metropolitan Lagos   6.537850   
141            Shomolu Pedro     100233     Metropolitan Lagos   6.537850   
160                Irepodun      104102  Non-Metropolitan Lagos  6.683795   

     Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
3     3.392800             1.0  Fast Food Restaurant           Bus Station   
7     3.286940             1.0        Clothing Store              Bus Stop   
8     3.312620             1.0  Fast Food Restaurant           Bus Station   
12    3.389040             1.0  Fast Food Restaurant           Bus Station   
14    3.223220             1.0     Convenience Store             Wine Shop   
15   

In [78]:
lagos_merged.loc[lagos_merged['Cluster Labels'] == 2]

Neighborhood  Postcode                 Area   Latitude  Longitude  \
11         Ajah     101245  Metropolitan Lagos   6.434441   3.589193   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11             2.0                 Beach             Wine Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11                 Diner           Fish Market                 Field   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11  Fast Food Restaurant        Farmers Market     Electronics Store   

   9th Most Common Venue 10th Most Common Venue  
11               Dog Run           Dessert Shop

In [79]:
lagos_merged.loc[lagos_merged['Cluster Labels'] == 3]

Neighborhood  Postcode                    Area   \
0                           Abarenji     100269     Metropolitan Lagos    
2                       Abule Ijesha     100213     Metropolitan Lagos    
4                      Adeniji Adele     101226     Metropolitan Lagos    
5                 Adeniran Ogunsanya     101211     Metropolitan Lagos    
6                            Agarawu     101229     Metropolitan Lagos    
10                           Ahmadia     100331     Metropolitan Lagos    
13                 Ajegunle Boundary     102211     Metropolitan Lagos    
21                             Allen     100281     Metropolitan Lagos    
27                     Apapa Central     101251     Metropolitan Lagos    
28                       Apapa North     101252     Metropolitan Lagos    
29                       Apapa South     101253     Metropolitan Lagos    
30                        Apapa West     101254     Metropolitan Lagos    
32                    Ayodele Street     100233     Metropolitan Lagos    
37                      Broad Street     100221     Metropolitan Lagos    
39                           Captain     100341     Metropolitan Lagos    
43                   Ebute Meta West     101211     Metropolitan Lagos    
46                            Egbeda     100276     Metropolitan Lagos    
52                           Gbosere     101231     Metropolitan Lagos    
63                           Ijedodo     100269     Metropolitan Lagos    
64                           Ijegemo     100269     Metropolitan Lagos    
67               Ijora Badia Central     102222     Metropolitan Lagos    
68                  Ijora Badia East     102221     Metropolitan Lagos    
69                 Ijora Badia North     102224     Metropolitan Lagos    
70                  Ijora Badia West     102223     Metropolitan Lagos    
71                       Ijora Oloye     101255     Metropolitan Lagos    
88                             Itire     100281     Metropolitan Lagos    
99                  Makoko Extension     101244     Metropolitan Lagos    
102                            Massy     101225     Metropolitan Lagos    
104                            Mende     100211     Metropolitan Lagos    
108                        Obanikoro     100232     Metropolitan Lagos    
116                          Okepopo     101224     Metropolitan Lagos    
118                      Olodi Apapa     102241     Metropolitan Lagos    
121                       Olowogbowo     101227     Metropolitan Lagos    
124                            Onike     100213     Metropolitan Lagos    
127                           Oregun     100212     Metropolitan Lagos    
128                    Oremeji Ifako     100241     Metropolitan Lagos    
129                      Orile Igamu     101271     Metropolitan Lagos    
132                       Oya Estate     100274     Metropolitan Lagos    
133                        Palmgrove     100252     Metropolitan Lagos    
134          Sari Iganmu Orile North     102232     Metropolitan Lagos    
135          Sari Iganmu Orile South     102231     Metropolitan Lagos    
136                          Sangisa     100248     Metropolitan Lagos    
138                           Site C     102263     Metropolitan Lagos    
143                 Surulere Central     101283     Metropolitan Lagos    
144                          Tujusho     101211     Metropolitan Lagos    
146            Yaba/ Ebute Meta East     101212     Metropolitan Lagos    
147                         Badagry      103101  Non-Metropolitan Lagos   
148  Ajara Agamathem Vetho/ Ibereko      103211  Non-Metropolitan Lagos   
150                           Ascon      103242  Non-Metropolitan Lagos   
151                 Bereks/Surulere      103251  Non-Metropolitan Lagos   
152                    Seme Badagry      103261  Non-Metropolitan Lagos   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0    6.454700   3.388760             3.0  Fast Food Resta

In [80]:
lagos_merged.loc[lagos_merged['Cluster Labels'] == 4]

Neighborhood  Postcode                    Area   Latitude  \
1                  Abule Egba     100276     Metropolitan Lagos   6.647500   
17            Alakija-Old Ojo     102262     Metropolitan Lagos   6.456280   
26                    Anthony     100232     Metropolitan Lagos   6.680581   
33                     Badore     100269     Metropolitan Lagos   6.497836   
34                Bakare Faro     102216     Metropolitan Lagos   6.706672   
38            Bungalow Estate     100264     Metropolitan Lagos   6.532150   
40             Dolphin Estate     101222     Metropolitan Lagos   6.459230   
44                       Egan     100268     Metropolitan Lagos   6.548620   
49        Festac Community II     102312     Metropolitan Lagos   6.473517   
50       Festac Community III     102313     Metropolitan Lagos   6.473517   
51        Festac Community IV     102314     Metropolitan Lagos   6.473517   
59                      Idumu     100276     Metropolitan Lagos   6.570320   
65                     Ijegun     100266     Metropolitan Lagos   6.516360   
66                 Ijeshatedo     101282     Metropolitan Lagos   6.491960   
75            Ikeja Oba Akran     100282     Metropolitan Lagos   6.607760   
78                      Ikoyi     101233     Metropolitan Lagos   6.456060   
81                      Ipaja     100278     Metropolitan Lagos   6.608830   
82                  Isale Eko     101226     Metropolitan Lagos   6.460884   
83                Isheri Ofin     100214     Metropolitan Lagos   6.575280   
85                Isheri Osun     100266     Metropolitan Lagos   6.515210   
87                    Itafaji     101225     Metropolitan Lagos   6.457215   
89                    Jankara     100312     Metropolitan Lagos   6.592570   
90                     Jibowu     100252     Metropolitan Lagos   6.659440   
96                    Lafiaji     101223     Metropolitan Lagos   6.457481   
98                     Magodo     100248     Metropolitan Lagos   6.630560   
100                    Marina     100221     Metropolitan Lagos   6.439100   
101                  Maryland     100211     Metropolitan Lagos   6.571110   
103                    Meiran     100361     Metropolitan Lagos   6.663609   
105  Murtala Muhammed Airport     100272     Metropolitan Lagos   6.570150   
107                  Obalende     101232     Metropolitan Lagos   6.446940   
111                     Ogudu     100242     Metropolitan Lagos   6.569470   
113                   Ojokoro     100313     Metropolitan Lagos   6.634580   
120                  Olowoira     100214     Metropolitan Lagos   6.629440   
122         Olute / Navy Town     102341     Metropolitan Lagos   6.453680   
126                     Opebi     100281     Metropolitan Lagos   6.591898   
130                    Oshodi     100261     Metropolitan Lagos   6.558170   
137             Satelite Town     102361     Metropolitan Lagos   6.452836   
139                  Shogunle     100261     Metropolitan Lagos   6.578060   
142                Suberu Oje     100351     Metropolitan Lagos   6.665336   
145           Victoria Island     101241     Metropolitan Lagos   6.428060   
149              Asago Sango      103241  Non-Metropolitan Lagos  6.465664   
156                    Erodo      106103  Non-Metropolitan Lagos  6.640680   
158                     Lekki     105102  Non-Metropolitan Lagos  6.409480   
159     Ikorodu rural areas       104101  Non-Metropolitan Lagos  6.623560   

     Longitude  Cluster Labels  1st Most Common Venue 2nd Most Common Venue  \
1     3.303060             4.0             Hookah Bar        Farmers Market   
17    3.270890             4.0                  Hotel         Hot Dog Joint   
26    3.356891             4.0              Gift Shop      Department Store   
33    3.585540             4.0                    Bar           Beer Garden   
34    3.332629             4.0            Beer Garden             Wine Shop   
38    3.298310             4.0  

# ACCRA 

In [84]:
accra_data = pd.read_excel(r"file:///C:\Users\User\Downloads\Ghana%20Data.xlsx") 
accra_data

Neighborhoods       Area 
0   Victoriaborg    Planned 
1     East Ridge    Planned 
2     West Ridge    Planned 
3    North Ridge    Planned 
4       Adabraka    Planned 
..           ...         ...
57     Mantseman  Unplanned 
58  Abossey Okai  Unplanned 
59       Apenkwa  Unplanned 
60     Bubiashie  Unplanned 
61        Mpoase  Unplanned 

[62 rows x 2 columns]

In [87]:
# define a function to get coordinates
def get_latlng(neighborhood):
    # initialize your variable to None
    lat_lng_coord = None
    # loop until you get the coordinates
    while(lat_lng_coord is None):
        g = geocoder.arcgis('{}, Accra, Ghana'.format(neighborhood))
        lat_lng_coord = g.latlng
    return lat_lng_coord

In [89]:
# call the function to get the coordinates, store in a new list using list comprehension
coord = [ get_latlng(neighborhood) for neighborhood in accra_data["Neighborhoods"].tolist() ]

In [90]:
coord

[[5.549280000000067, -0.19714999999996508],
 [5.549940038728501, -0.20013495357713396],
 [5.549940038728501, -0.20013495357713396],
 [5.572300000000041, -0.19618999999994458],
 [5.5788800000000265, -0.23260999999996557],
 [5.569300000000055, -0.20623999999997977],
 [5.542570000000069, -0.2631699999999455],
 [5.612405040586552, -0.17399489887709763],
 [5.600080000000048, -0.17583999999993694],
 [5.549940038728501, -0.20013495357713396],
 [5.577014888866524, -0.19057591722192693],
 [5.609377215682804, -0.2035244350218854],
 [5.647708509348373, -0.17975750689350165],
 [5.562970000000064, -0.24105999999994765],
 [5.565920000000062, -0.21651999999994587],
 [5.599860000000035, -0.21619999999995798],
 [5.647708509348373, -0.17975750689350165],
 [5.60618621818487, -0.2221855605555236],
 [5.577212054530421, -0.18049992629349276],
 [5.568666973416073, -0.17249188604054166],
 [5.58348011482042, -0.14930790372481811],
 [5.5788800000000265, -0.23260999999996557],
 [5.536540000000059, -0.24182999999

In [91]:
df_coord = pd.DataFrame(coord, columns=['Latitude', 'Longitude'])

In [92]:
# merge the coordinates into the original dataframe
accra_data['Latitude'] = df_coord['Latitude']
accra_data['Longitude'] = df_coord['Longitude']

In [93]:
print(accra_data.shape)
accra_data.head(12)

(62, 4)


Neighborhoods     Area   Latitude  Longitude
0               Victoriaborg  Planned   5.549280  -0.197150
1                 East Ridge  Planned   5.549940  -0.200135
2                 West Ridge  Planned   5.549940  -0.200135
3                North Ridge  Planned   5.572300  -0.196190
4                   Adabraka  Planned   5.578880  -0.232610
5                Asylum Down  Planned   5.569300  -0.206240
6              McCarthy Hill  Planned   5.542570  -0.263170
7              Airport Hills  Planned   5.612405  -0.173995
8   Airport Residential Area  Planned   5.600080  -0.175840
9                Roman Ridge  Planned   5.549940  -0.200135
10                     Kanda  Planned   5.577015  -0.190576
11                  Dzorwulu  Planned   5.609377  -0.203524

In [94]:
address = 'Accra, Ghana'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Accra are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Accra are 5.5600141, -0.2057437.


In [95]:
map_accra = folium.Map(location=[latitude, longitude], zoom_start=11)

In [96]:
# add markers to map
for lat, lng, label in zip(accra_data['Latitude'], accra_data['Longitude'], accra_data['Neighborhoods']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_accra)  
    
map_accra

In [97]:
accra_data

Neighborhoods       Area   Latitude  Longitude
0   Victoriaborg    Planned   5.549280  -0.197150
1     East Ridge    Planned   5.549940  -0.200135
2     West Ridge    Planned   5.549940  -0.200135
3    North Ridge    Planned   5.572300  -0.196190
4       Adabraka    Planned   5.578880  -0.232610
..           ...         ...       ...        ...
57     Mantseman  Unplanned   5.213030  -1.062310
58  Abossey Okai  Unplanned   5.553380  -0.223934
59       Apenkwa  Unplanned   5.600430  -0.233120
60     Bubiashie  Unplanned   5.574561  -0.248072
61        Mpoase  Unplanned   5.542343  -0.243245

[62 rows x 4 columns]

In [98]:
radius = 2000
LIMIT = 100

venues = []

for lat, long, neighborhood in zip(accra_data['Latitude'], accra_data['Longitude'], accra_data['Neighborhoods']):
    
    # create the API request URL
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    
    # make the GET request
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    # return only relevant information for each nearby venue
    for venue in results:
        venues.append((
            neighborhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

In [99]:
# convert the venues list into a new DataFrame
venues_df = pd.DataFrame(venues)

# define the column names
venues_df.columns = ['Neighborhoods', 'Latitude', 'Longitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']

print(venues_df.shape)
venues_df

(1953, 7)


Neighborhoods  Latitude  Longitude  \
0     Victoriaborg  5.549280  -0.197150   
1     Victoriaborg  5.549280  -0.197150   
2     Victoriaborg  5.549280  -0.197150   
3     Victoriaborg  5.549280  -0.197150   
4     Victoriaborg  5.549280  -0.197150   
...            ...       ...        ...   
1948        Mpoase  5.542343  -0.243245   
1949        Mpoase  5.542343  -0.243245   
1950        Mpoase  5.542343  -0.243245   
1951        Mpoase  5.542343  -0.243245   
1952        Mpoase  5.542343  -0.243245   

                                           VenueName  VenueLatitude  \
0                    Kempinski Hotel Gold Coast City       5.554440   
1                         Mövenpick Ambassador Hotel       5.554508   
2                                  Black Star Square       5.548006   
3     The Deli Restaurant Movenpick Ambassador Hotel       5.553812   
4                         Accra City Hotel (Novotel)       5.554128   
...                                              ...            ...   
1948                   Young Star Boutique ( Y.S.B )       5.558471   
1949                                   Club Infiniti       5.539580   
1950                                      Tripplesea       5.535291   
1951                              Rocky's collection       5.558458   
1952        The Shoppe (Korlebu Welfare Supermarket)       5.535958   

      VenueLongitude      VenueCategory  
0          -0.197899              Hotel  
1          -0.202435              Hotel  
2          -0.192508      Historic Site  
3          -0.202396         Restaurant  
4          -0.203456              Hotel  
...              ...                ...  
1948       -0.244172           Boutique  
1949       -0.244548          Nightclub  
1950       -0.231032  Electronics Store  
1951       -0.244181           Boutique  
1952       -0.227299   Department Store  

[1953 rows x 7 columns]

In [100]:
venues_df.groupby(["Neighborhoods"]).count()

Latitude  Longitude  VenueName  VenueLatitude  \
Neighborhoods                                                         
Abeka                       27         27         27             27   
Abelemkpe                   19         19         19             19   
Abossey Okai                25         25         25             25   
Abuja                       32         32         32             32   
Accra New Town              28         28         28             28   
...                        ...        ...        ...            ...   
Tesano                      36         36         36             36   
Teshie                      84         84         84             84   
Victoriaborg                38         38         38             38   
West Legon/Westlands        27         27         27             27   
West Ridge                  30         30         30             30   

                      VenueLongitude  VenueCategory  
Neighborhoods                                        
Abeka                             27             27  
Abelemkpe                         19             19  
Abossey Okai                      25             25  
Abuja                             32             32  
Accra New Town                    28             28  
...                              ...            ...  
Tesano                            36             36  
Teshie                            84             84  
Victoriaborg                      38             38  
West Legon/Westlands              27             27  
West Ridge                        30             30  

[61 rows x 6 columns]

In [101]:
# print out the list of categories
venues_df['VenueCategory'].unique()[:60]

array(['Hotel', 'Historic Site', 'Restaurant', 'Pool',
       'American Restaurant', 'Swiss Restaurant', 'African Restaurant',
       'Bar', 'Snack Place', 'Theater', 'Plaza', 'Indian Restaurant',
       'History Museum', 'Coffee Shop', 'Chinese Restaurant',
       'Concert Hall', 'Soccer Field', 'Shopping Mall', 'Park',
       'Rest Area', 'Market', 'Hotel Bar', 'Flea Market', 'Pharmacy',
       'Convenience Store', 'Clothing Store', 'Mobile Phone Shop',
       'Night Market', 'Paper / Office Supplies Store',
       'Electronics Store', 'BBQ Joint', 'Nightclub', 'Dessert Shop',
       'Jazz Club', 'Music Venue', 'Modern European Restaurant', 'Pub',
       'Sports Bar', 'Pizza Place', 'Spanish Restaurant',
       'Italian Restaurant', 'Cocktail Bar', 'Hostel', 'Department Store',
       'Lounge', 'Thai Restaurant', 'Ice Cream Shop',
       'Fast Food Restaurant', 'Bakery', 'Bed & Breakfast',
       'Kebab Restaurant', 'Food Court', 'Luggage Store', 'Supermarket',
       'Grocery Store'

In [102]:
print('There are {} uniques categories.'.format(len(venues_df['VenueCategory'].unique())))

There are 133 uniques categories.


In [103]:
"Neighborhoods" in venues_df['VenueCategory'].unique()

False

In [104]:
# one hot encoding
accra_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
accra_onehot['Neighborhoods'] = venues_df['Neighborhoods'] 

# move neighborhood column to the first column
fixed_columns = [accra_onehot.columns[-1]] + list(accra_onehot.columns[:-1])
accra_onehot = accra_onehot[fixed_columns]

print(accra_onehot.shape)
accra_onehot.head()

(1953, 134)


Neighborhoods  African Restaurant  Airport  Airport Lounge  \
0  Victoriaborg                   0        0               0   
1  Victoriaborg                   0        0               0   
2  Victoriaborg                   0        0               0   
3  Victoriaborg                   0        0               0   
4  Victoriaborg                   0        0               0   

   Airport Terminal  American Restaurant  Art Gallery  Arts & Crafts Store  \
0                 0                    0            0                    0   
1                 0                    0            0                    0   
2                 0                    0            0                    0   
3                 0                    0            0                    0   
4                 0                    0            0                    0   

   Arts & Entertainment  Asian Restaurant  ...  Spanish Restaurant  \
0                     0                 0  ...                   0   
1                     0                 0  ...                   0   
2                     0                 0  ...                   0   
3                     0                 0  ...                   0   
4                     0                 0  ...                   0   

   Sports Bar  Supermarket  Swiss Restaurant  Thai Restaurant  Theater  \
0           0            0                 0                0        0   
1           0            0                 0                0        0   
2           0            0                 0                0        0   
3           0            0                 0                0        0   
4           0            0                 0                0        0   

   Theme Park  Trail  Turkish Restaurant  Wings Joint  
0           0      0                   0            0  
1           0      0                   0            0  
2           0      0                   0            0  
3           0      0                   0            0  
4           0      0                   0            0  

[5 rows x 134 columns]

In [105]:
accra_grouped = accra_onehot.groupby(["Neighborhoods"]).mean().reset_index()

print(accra_grouped.shape)
accra_grouped

(61, 134)


Neighborhoods  African Restaurant   Airport  Airport Lounge  \
0                  Abeka            0.074074  0.000000             0.0   
1              Abelemkpe            0.052632  0.000000             0.0   
2           Abossey Okai            0.000000  0.000000             0.0   
3                  Abuja            0.000000  0.000000             0.0   
4         Accra New Town            0.035714  0.000000             0.0   
..                   ...                 ...       ...             ...   
56                Tesano            0.027778  0.000000             0.0   
57                Teshie            0.083333  0.011905             0.0   
58          Victoriaborg            0.052632  0.000000             0.0   
59  West Legon/Westlands            0.074074  0.000000             0.0   
60            West Ridge            0.033333  0.000000             0.0   

    Airport Terminal  American Restaurant  Art Gallery  Arts & Crafts Store  \
0               0.00             0.000000     0.000000                  0.0   
1               0.00             0.052632     0.000000                  0.0   
2               0.04             0.000000     0.000000                  0.0   
3               0.00             0.000000     0.000000                  0.0   
4               0.00             0.000000     0.000000                  0.0   
..               ...                  ...          ...                  ...   
56              0.00             0.000000     0.027778                  0.0   
57              0.00             0.023810     0.011905                  0.0   
58              0.00             0.026316     0.000000                  0.0   
59              0.00             0.000000     0.037037                  0.0   
60              0.00             0.000000     0.000000                  0.0   

    Arts & Entertainment  Asian Restaurant  ...  Spanish Restaurant  \
0               0.000000          0.000000  ...                 0.0   
1               0.000000          0.000000  ...                 0.0   
2               0.000000          0.040000  ...                 0.0   
3               0.000000          0.000000  ...                 0.0   
4               0.000000          0.035714  ...                 0.0   
..                   ...               ...  ...                 ...   
56              0.000000          0.000000  ...                 0.0   
57              0.011905          0.000000  ...                 0.0   
58              0.000000          0.000000  ...                 0.0   
59              0.000000          0.000000  ...                 0.0   
60              0.000000          0.000000  ...                 0.0   

    Sports Bar  Supermarket  Swiss Restaurant  Thai Restaurant   Theater  \
0     0.000000     0.000000          0.000000         0.000000  0.000000   
1     0.000000     0.000000          0.000000         0.000000  0.000000   
2     0.000000     0.040000          0.000000         0.000000  0.000000   
3     0.000000     0.000000          0.000000         0.000000  0.000000   
4     0.035714     0.000000          0.000000         0.000000  0.000000   
..         ...          ...               ...              ...       ...   
56    0.000000     0.000000          0.000000         0.000000  0.000000   
57    0.000000     0.011905          0.000000         0.011905  0.000000   
58    0.000000     0.000000          0.026316         0.000000  0.026316   
59    0.000000     0.000000          0.000000         0.000000  0.000000   
60    0.000000     0.000000          0.033333         0.000000  0.033333   

    Theme Park  Trail  Turkish Restaurant  Wings Joint  
0          0.0    0.0            0.000000          0.0  
1          0.0    0.0            0.000000          0.0  
2          0.0    0.0            0.000000          0.0  
3          0.0    0.0            0.000000          0.0  
4          0.0    0.0            0.000000          0.0  
..         ...    ...                 ...          ...  
56      

In [106]:
accra_grouped.shape

(61, 134)

In [107]:
accra_grouped = accra_onehot.groupby('Neighborhoods').mean().reset_index()
accra_grouped

Neighborhoods  African Restaurant   Airport  Airport Lounge  \
0                  Abeka            0.074074  0.000000             0.0   
1              Abelemkpe            0.052632  0.000000             0.0   
2           Abossey Okai            0.000000  0.000000             0.0   
3                  Abuja            0.000000  0.000000             0.0   
4         Accra New Town            0.035714  0.000000             0.0   
..                   ...                 ...       ...             ...   
56                Tesano            0.027778  0.000000             0.0   
57                Teshie            0.083333  0.011905             0.0   
58          Victoriaborg            0.052632  0.000000             0.0   
59  West Legon/Westlands            0.074074  0.000000             0.0   
60            West Ridge            0.033333  0.000000             0.0   

    Airport Terminal  American Restaurant  Art Gallery  Arts & Crafts Store  \
0               0.00             0.000000     0.000000                  0.0   
1               0.00             0.052632     0.000000                  0.0   
2               0.04             0.000000     0.000000                  0.0   
3               0.00             0.000000     0.000000                  0.0   
4               0.00             0.000000     0.000000                  0.0   
..               ...                  ...          ...                  ...   
56              0.00             0.000000     0.027778                  0.0   
57              0.00             0.023810     0.011905                  0.0   
58              0.00             0.026316     0.000000                  0.0   
59              0.00             0.000000     0.037037                  0.0   
60              0.00             0.000000     0.000000                  0.0   

    Arts & Entertainment  Asian Restaurant  ...  Spanish Restaurant  \
0               0.000000          0.000000  ...                 0.0   
1               0.000000          0.000000  ...                 0.0   
2               0.000000          0.040000  ...                 0.0   
3               0.000000          0.000000  ...                 0.0   
4               0.000000          0.035714  ...                 0.0   
..                   ...               ...  ...                 ...   
56              0.000000          0.000000  ...                 0.0   
57              0.011905          0.000000  ...                 0.0   
58              0.000000          0.000000  ...                 0.0   
59              0.000000          0.000000  ...                 0.0   
60              0.000000          0.000000  ...                 0.0   

    Sports Bar  Supermarket  Swiss Restaurant  Thai Restaurant   Theater  \
0     0.000000     0.000000          0.000000         0.000000  0.000000   
1     0.000000     0.000000          0.000000         0.000000  0.000000   
2     0.000000     0.040000          0.000000         0.000000  0.000000   
3     0.000000     0.000000          0.000000         0.000000  0.000000   
4     0.035714     0.000000          0.000000         0.000000  0.000000   
..         ...          ...               ...              ...       ...   
56    0.000000     0.000000          0.000000         0.000000  0.000000   
57    0.000000     0.011905          0.000000         0.011905  0.000000   
58    0.000000     0.000000          0.026316         0.000000  0.026316   
59    0.000000     0.000000          0.000000         0.000000  0.000000   
60    0.000000     0.000000          0.033333         0.000000  0.033333   

    Theme Park  Trail  Turkish Restaurant  Wings Joint  
0          0.0    0.0            0.000000          0.0  
1          0.0    0.0            0.000000          0.0  
2          0.0    0.0            0.000000          0.0  
3          0.0    0.0            0.000000          0.0  
4          0.0    0.0            0.000000          0.0  
..         ...    ...                 ...          ...  
56      

In [108]:
num_top_venues = 5

for hood in accra_grouped['Neighborhoods']:
    print("----" +hood+ "----")
    temp = accra_grouped[accra_grouped['Neighborhoods'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Abeka----
                venue  freq
0         Bus Station  0.11
1  African Restaurant  0.07
2                 Pub  0.07
3  Chinese Restaurant  0.07
4               Hotel  0.07


----Abelemkpe----
                  venue  freq
0  Fast Food Restaurant  0.16
1                 Hotel  0.11
2           Bus Station  0.05
3           Snack Place  0.05
4            Restaurant  0.05


----Abossey Okai----
               venue  freq
0        Bus Station  0.32
1  Mobile Phone Shop  0.08
2             Casino  0.08
3        Supermarket  0.04
4   Department Store  0.04


----Abuja----
                  venue  freq
0           Bus Station  0.19
1          Soccer Field  0.06
2  Fast Food Restaurant  0.06
3          Cocktail Bar  0.03
4              Pharmacy  0.03


----Accra New Town----
                  venue  freq
0  Fast Food Restaurant  0.11
1         Shopping Mall  0.07
2             Nightclub  0.07
3                   Bar  0.07
4    African Restaurant  0.04


----Achimota----
           ve

In [109]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [110]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhoods']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhoods'] = accra_grouped['Neighborhoods']

for ind in np.arange(accra_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(accra_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhoods 1st Most Common Venue 2nd Most Common Venue  \
0                  Abeka           Bus Station                 Hotel   
1              Abelemkpe  Fast Food Restaurant                 Hotel   
2           Abossey Okai           Bus Station                Casino   
3                  Abuja           Bus Station          Soccer Field   
4         Accra New Town  Fast Food Restaurant             Nightclub   
..                   ...                   ...                   ...   
56                Tesano  Fast Food Restaurant              Pharmacy   
57                Teshie    African Restaurant                Bakery   
58          Victoriaborg                 Hotel    Chinese Restaurant   
59  West Legon/Westlands  Fast Food Restaurant                   Pub   
60            West Ridge                 Hotel    Chinese Restaurant   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                    Pub    Chinese Restaurant    African Restaurant   
1             Restaurant   American Restaurant           Bus Station   
2      Mobile Phone Shop      Department Store      Sculpture Garden   
3   Fast Food Restaurant                 Hotel        Cosmetics Shop   
4                    Bar         Shopping Mall                 Hotel   
..                   ...                   ...                   ...   
56                   Bar            Food Court                 Hotel   
57    Chinese Restaurant  Fast Food Restaurant                Lounge   
58     Indian Restaurant    African Restaurant               Theater   
59                 Hotel            Food Court    African Restaurant   
60            Restaurant     Indian Restaurant                Market   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0               Restaurant                  Food           Flea Market   
1                     Café    Chinese Restaurant        Cosmetics Shop   
2              Supermarket           Flea Market     Indian Restaurant   
3                     Café                Casino              Pharmacy   
4   Furniture / Home Store           Music Venue         Luggage Store   
..                     ...                   ...                   ...   
56               Nightclub           Music Venue         Movie Theater   
57                     Bar                  Café     Electronics Store   
58             Coffee Shop                  Pool                 Plaza   
59               Multiplex            Smoke Shop           Pizza Place   
60                    Pool                 Plaza              Pharmacy   

   9th Most Common Venue 10th Most Common Venue  
0         Cosmetics Shop   Fast Food Restaurant  
1                  Diner             Food Court  
2       Asian Restaurant     Athletics & Sports  
3     Chinese Restaurant              Nightclub  
4              Jazz Club              Hotel Bar  
..                   ...                    ...  
56     Indian Restaurant          Grocery Store  
57                 Hotel              BBQ Joint  
58              Pharmacy                   Park  
59         Luggage Store       Basketball Court  
60                  Park              Nightclub  

[61 rows x 11 columns]

### Clustering the Neighborhoods Of Accra 

In [113]:
# set number of clusters
kclusters = 5

accra_grouped_clustering = accra_grouped.drop('Neighborhoods', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(accra_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 0, 0, 4, 4, 0, 4, 2, 2])

In [114]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
accra_merged = accra_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
accra_merged = accra_merged.join(neighborhoods_venues_sorted.set_index('Neighborhoods'), on='Neighborhoods')

accra_merged.head() # check the last columns!

Neighborhoods     Area   Latitude  Longitude  Cluster Labels  \
0  Victoriaborg  Planned    5.54928  -0.197150               2   
1    East Ridge  Planned    5.54994  -0.200135               2   
2    West Ridge  Planned    5.54994  -0.200135               2   
3   North Ridge  Planned    5.57230  -0.196190               4   
4      Adabraka  Planned    5.57888  -0.232610               0   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                 Hotel    Chinese Restaurant     Indian Restaurant   
1                 Hotel    Chinese Restaurant            Restaurant   
2                 Hotel    Chinese Restaurant            Restaurant   
3    African Restaurant    Chinese Restaurant                 Hotel   
4           Bus Station          Soccer Field  Fast Food Restaurant   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0    African Restaurant               Theater           Coffee Shop   
1     Indian Restaurant                Market                  Pool   
2     Indian Restaurant                Market                  Pool   
3  Fast Food Restaurant                   Bar           Pizza Place   
4                 Hotel        Cosmetics Shop                  Café   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                  Pool                 Plaza              Pharmacy   
1                 Plaza              Pharmacy                  Park   
2                 Plaza              Pharmacy                  Park   
3                   Pub             Nightclub        Clothing Store   
4                Casino              Pharmacy    Chinese Restaurant   

  10th Most Common Venue  
0                   Park  
1              Nightclub  
2              Nightclub  
3                 Bakery  
4              Nightclub

In [115]:
accra_merged = accra_merged.dropna()

In [116]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(accra_merged['Latitude'], accra_merged['Longitude'], accra_merged['Neighborhoods'], accra_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [117]:
accra_merged.loc[accra_merged['Cluster Labels'] == 0]

Neighborhoods       Area   Latitude  Longitude  \
4                          Adabraka    Planned   5.578880  -0.232610   
13                         Kaneshie    Planned   5.562970  -0.241060   
21                          Chorkor    Planned   5.578880  -0.232610   
24                      Korle Gonno    Planned   5.532700  -0.226680   
26                         Mataheko    Planned   5.559118  -0.235135   
31  Old Fadama (Sodom and Gomorrah)      Slums   5.578880  -0.232610   
33                            Abuja      Slums   5.578880  -0.232610   
34                          Chemuna      Slums   5.578880  -0.232610   
35                       Gbegbeyise      Slums   5.578880  -0.232610   
36                          Chorkor      Slums   5.578880  -0.232610   
40                          Awoshie  Unplanned   5.586001  -0.277186   
50                            Lapaz  Unplanned   5.578880  -0.232610   
51                         Darkuman  Unplanned   5.594010  -0.252560   
55                     Nii Boi Town  Unplanned   5.578880  -0.232610   
58                     Abossey Okai  Unplanned   5.553380  -0.223934   
60                        Bubiashie  Unplanned   5.574561  -0.248072   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
4                0           Bus Station          Soccer Field   
13               0           Bus Station                 Hotel   
21               0           Bus Station          Soccer Field   
24               0      Department Store     Electronics Store   
26               0              Boutique           Bus Station   
31               0           Bus Station          Soccer Field   
33               0           Bus Station          Soccer Field   
34               0           Bus Station          Soccer Field   
35               0           Bus Station          Soccer Field   
36               0           Bus Station          Soccer Field   
40               0           Bus Station  Fast Food Restaurant   
50               0           Bus Station          Soccer Field   
51               0           Bus Station                 Hotel   
55               0           Bus Station          Soccer Field   
58               0           Bus Station                Casino   
60               0           Bus Station            Restaurant   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
4   Fast Food Restaurant                 Hotel        Cosmetics Shop   
13              Boutique          Soccer Field            Restaurant   
21  Fast Food Restaurant                 Hotel        Cosmetics Shop   
24            Lighthouse           Bus Station                  Food   
26     Mobile Phone Shop          Soccer Field      Sculpture Garden   
31  Fast Food Restaurant                 Hotel        Cosmetics Shop   
33  Fast Food Restaurant                 Hotel        Cosmetics Shop   
34  Fast Food Restaurant                 Hotel        Cosmetics Shop   
35  Fast Food Restaurant                 Hotel        Cosmetics Shop   
36  Fast Food Restaurant                 Hotel        Cosmetics Shop   
40                   Bar                Hostel             Nightclub   
50  Fast Food Restaurant                 Hotel        Cosmetics Shop   
51                  Food                   Gym                   Pub   
55  Fast Food Restaurant                 Hotel        Cosmetics Shop   
58     Mobile Phone Shop      Department Store      Sculpture Garden   
60              Boutique  Fast Food Restaurant                 Hotel   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
4                   Café                Casino              Pharmacy   
13          Noodle House          Home Service      Sculpture Garden   
21                  Café                Casino              Pharmacy   
24      Football Stadium            Food Truck            Food Court   
26           Snack Place           Supermarket     Fish & Chips Shop   
31               

In [118]:
accra_merged.loc[accra_merged['Cluster Labels'] == 1]

Neighborhoods       Area   Latitude  Longitude  Cluster Labels  \
22      Mamprobi    Planned    5.53654   -0.24183               1   
56    Akweteyman  Unplanned    5.51667   -0.26667               1   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
22    African Restaurant                   Pub         Movie Theater   
56    African Restaurant  Fast Food Restaurant     French Restaurant   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
22      Department Store  Fast Food Restaurant             Nightclub   
56          Dessert Shop                 Diner     Electronics Store   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
22     Electronics Store    Falafel Restaurant        Farmers Market   
56    Falafel Restaurant        Farmers Market     Fish & Chips Shop   

   10th Most Common Venue  
22                  Diner  
56            Flea Market

In [119]:
accra_merged.loc[accra_merged['Cluster Labels'] == 2]

Neighborhoods       Area   Latitude  Longitude  \
0                         Victoriaborg    Planned   5.549280  -0.197150   
1                           East Ridge    Planned   5.549940  -0.200135   
2                           West Ridge    Planned   5.549940  -0.200135   
5                          Asylum Down    Planned   5.569300  -0.206240   
7                        Airport Hills    Planned   5.612405  -0.173995   
8             Airport Residential Area    Planned   5.600080  -0.175840   
9                          Roman Ridge    Planned   5.549940  -0.200135   
11                            Dzorwulu    Planned   5.609377  -0.203524   
17                           Abelemkpe    Planned   5.606186  -0.222186   
20  Burma Camp (formerly Giffard Camp)    Planned   5.583480  -0.149308   
29                       Agbogbloshie]      Slums   5.551286  -0.214114   
30                         Sabon Zongo      Slums   5.544156  -0.210656   
32               Lavender Hill (Accra)      Slums   5.643556  -0.155873   
37                        Agbogbloshie  Unplanned   5.551286  -0.214114   
43                            Kotobabi  Unplanned   5.595940  -0.200589   
46                            Pig Farm  Unplanned   5.598357  -0.200628   
48                             Maamobi  Unplanned   5.592953  -0.192368   
49                               Abeka  Unplanned   5.602603  -0.241608   
59                             Apenkwa  Unplanned   5.600430  -0.233120   

    Cluster Labels 1st Most Common Venue   2nd Most Common Venue  \
0                2                 Hotel      Chinese Restaurant   
1                2                 Hotel      Chinese Restaurant   
2                2                 Hotel      Chinese Restaurant   
5                2                 Hotel               Hotel Bar   
7                2                 Hotel              Restaurant   
8                2                 Hotel              Restaurant   
9                2                 Hotel      Chinese Restaurant   
11               2                 Hotel               Nightclub   
17               2  Fast Food Restaurant                   Hotel   
20               2              Pharmacy                   Hotel   
29               2                 Hotel  Furniture / Home Store   
30               2                 Hotel          Clothing Store   
32               2                 Hotel                     Pub   
37               2                 Hotel  Furniture / Home Store   
43               2                 Hotel            Soccer Field   
46               2                 Hotel            Soccer Field   
48               2                 Hotel            Soccer Field   
49               2           Bus Station                   Hotel   
59               2                 Hotel             Bus Station   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0      Indian Restaurant    African Restaurant               Theater   
1             Restaurant     Indian Restaurant                Market   
2             Restaurant     Indian Restaurant                Market   
5                    Bar          Cocktail Bar  Fast Food Restaurant   
7     Chinese Restaurant    African Restaurant     Electronics Store   
8     African Restaurant    Chinese Restaurant      Department Store   
9             Restaurant     Indian Restaurant                Market   
11    Chinese Restaurant            Restaurant  Gym / Fitness Center   
17            Restaurant   American Restaurant           Bus Station   
20     Convenience Store          Soccer Field         Military Base   
29           Bus Station                  Pool              Pharmacy   
30            Restaurant                Market            Lighthouse   
32         Shopping Mall                  Café      Department Store   
37           Bus Station                  Pool              Pharmacy   
43  Fast Food Restaurant  Gym / Fitness Center            Restaurant   
46  Fast Fo

In [120]:
accra_merged.loc[accra_merged['Cluster Labels'] == 3]

Neighborhoods       Area   Latitude  Longitude  Cluster Labels  \
57     Mantseman  Unplanned    5.21303   -1.06231               3   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
57                 Motel           Wings Joint     French Restaurant   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
57                 Diner     Electronics Store    Falafel Restaurant   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
57        Farmers Market  Fast Food Restaurant     Fish & Chips Shop   

   10th Most Common Venue  
57            Flea Market

In [121]:
accra_merged.loc[accra_merged['Cluster Labels'] == 4]

Neighborhoods       Area   Latitude  Longitude  Cluster Labels  \
3            North Ridge    Planned   5.572300  -0.196190               4   
6          McCarthy Hill    Planned   5.542570  -0.263170               4   
10                 Kanda    Planned   5.577015  -0.190576               4   
12            East Legon    Planned   5.647709  -0.179758               4   
14            Kokomlemle    Planned   5.565920  -0.216520               4   
15                Tesano    Planned   5.599860  -0.216200               4   
16  West Legon/Westlands    Planned   5.647709  -0.179758               4   
18           Cantonments    Planned   5.577212  -0.180500               4   
19                Labone    Planned   5.568667  -0.172492               4   
23              Korle Bu    Planned   5.555175  -0.235819               4   
25       Lartebiokorshie    Planned   5.553750  -0.240800               4   
27              Dansoman    Planned   5.559435  -0.260232               4   
28                Adenta    Planned   5.714170  -0.154180               4   
38                  Nima  Unplanned   5.583900  -0.199890               4   
39                 Kpehe  Unplanned   5.583330  -0.216690               4   
41                 Alajo  Unplanned   5.598587  -0.215221               4   
42                Avenor  Unplanned   5.579290  -0.216950               4   
44            Bawaleshie  Unplanned   5.642251  -0.163712               4   
45     Christian Village  Unplanned   5.633409  -0.216374               4   
47        Accra New Town  Unplanned   5.578820  -0.208890               4   
52              Achimota  Unplanned   5.615325  -0.210776               4   
53                Teshie  Unplanned   5.566581  -0.167889               4   
54               Odorkor  Unplanned   5.584825  -0.212644               4   
61                Mpoase  Unplanned   5.542343  -0.243245               4   

   1st Most Common Venue   2nd Most Common Venue   3rd Most Common Venue  \
3     African Restaurant      Chinese Restaurant                   Hotel   
6                    Bar            Burger Joint           Shopping Mall   
10    African Restaurant    Fast Food Restaurant                     Bar   
12  Fast Food Restaurant                     Pub                   Hotel   
14    African Restaurant                     Bar  Furniture / Home Store   
15  Fast Food Restaurant                Pharmacy                     Bar   
16  Fast Food Restaurant                     Pub                   Hotel   
18    African Restaurant      Chinese Restaurant                   Hotel   
19    African Restaurant      Chinese Restaurant                  Bakery   
23           Bus Station       Electronics Store                Boutique   
25              Boutique          Ice Cream Shop             Bus Station   
27  Fast Food Restaurant                     Bar              Restaurant   
28    African Restaurant                     Bar                Pharmacy   
38                 Hotel                     Bar               Nightclub   
39  Fast Food Restaurant                  Bakery                  Hostel   
41  Fast Food Restaurant                Pharmacy              Food Court   
42  Fast Food Restaurant  Furniture / Home Store           Shopping Mall   
44    African Restaurant                     Pub              Restaurant   
45   American Restaurant             Golf Course       Convenience Store   
47  Fast Food Restaurant               Nightclub                     Bar   
52           Golf Course             Bus Station              Food Court   
53    African Restaurant                  Bakery      Chinese Restaurant   
54  Fast Food Restaurant           Shopping Mall      African Restaurant   
61              Boutique       Electronics Store               Nightclub   

   4th Most Common Venue 5th Most Common Venue   6th Most Common Venue  \
3   Fast Food Restaurant                   Bar             Pizza Place   
6             Theme Park     Electro

### Result and Dicsussion 

We reached at the end of the analysis, From the cluster arrangement from the information above for the both cities Lagos and Accra we see that Lagos is a bigger commercial city than Accra, Lagos has more recreational centres and hangout spots like Hotels, fast food restaurants, Bar,  Nightclub and lounges. Also due to the closeness of lagos to the atlantic ocean there are presnece of Beaches in Lagos. Compared to Lagos, Accra is smaller with lesser activities we find hotels, fast food restaurants, bus station and african restaurants as the most popular venus in Accra. 

### Conclusion

In this project we have tried to analyze the dataset of two West African countries commercial city Lagos and Accra, we analyzed the neighborhoods in the regions and tried to cluster them based on location and popular places like restaurants, hotels and other top location and nearby venus. For the city of Lagos we explored the city based on their postal codes and then extracted the common venues present in each of the neighbourhoods finally concluding with clustering similar neighbourhoods together.
From the result we see that both cities are somewhat similar most common venues include Hotels, restaurant, fast food restaurants and bus stops. From the cluster above we see that Lagos has more hotels and fast food restaurants as compared to Accra. 